### ISSUE TITLE
The out put of the KAN model is a matrix fulled of (NaN) in second time iterations

### ISSUE THREAD

**Issue by linmi1** [KindXiaoming/pykan#230]
I replace the MLP of NeRF,But get the NaN loss because of the KAN model's output
initialize KAN:
`net = KAN(width=[84,1,4], grid=5, k=3, seed=0,device=device)`
``
#----KAN------
    device = pts_flat.device  
    pts_flat = pts_flat.to(device)
    dir_flat = dir_flat.to(device)
    data=torch.cat((pts_flat, dir_flat), dim=-1)
    model_out = net(data)
    **print (model_out)  # here is the print**
    rgb  = model_out[:,0:3]
    sigma = model_out[:,3]
# ------------------
``
``
here is the print :
Epoch 1:   0%|                                            | 1/976 [00:00<09:24,  1.73it/s, loss=inf]
tensor([[ 0.0486,  0.0511,  0.0563,  0.0445],
        [ 0.0033, -0.0027,  0.0009, -0.0026],
        [ 0.2292,  0.2569,  0.2867,  0.2423],
        ...,
        [-0.0786, -0.0964, -0.1001, -0.0851],
        [-0.0152, -0.0245, -0.0216, -0.0215],
        [ 0.1987,  0.2231,  0.2465,  0.2082]], device='cuda:0',
       grad_fn=<AddBackward0>)
Epoch 1:   0%|                                            | 2/976 [00:03<31:20,  1.93s/it, loss=nan]
tensor([[nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        ...,
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan]], device='cuda:0', grad_fn=<AddBackward0>)
Epoch 1:   0%|▏                                           | 3/976 [00:03<18:57,  1.17s/it, loss=nan]
tensor([[nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        ...,
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan]], device='cuda:0', grad_fn=<AddBackward0>)
Epoch 1:   0%|▏                                           | 4/976 [00:06<28:42,  1.77s/it, loss=nan]
tensor([[nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        ...,
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan]], device='cuda:0', grad_fn=<AddBackward0>)``

And after that I check the **backward code**:
``
for i in range(iterations):
            train_rays = next(train_iter)
            assert train_rays.shape == (Batch_size, 9)

            rays_o, rays_d, target_rgb = torch.chunk(train_rays, 3, dim=-1)
            rays_od = (rays_o, rays_d)
            rgb, _, __ = render_rays(net, rays_od, bound=bound, N_samples=N_samples, device=device, use_view=use_view)
            
#             print(rgb, target_rgb)
            acts_scale=[84,0.25]
            
            loss = mse(rgb, target_rgb)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            p_bar.set_postfix({'loss': '{0:1.5f}'.format(loss.item())})
            p_bar.update(1)
``
Maybe the way to calculation loss is wrong?
WHO can help me ?  THANK YOU!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by linmi1** [comment]
I find that the value of output of the KAN model maybe negative ,so the calculation of loss will be troubled.How to keep the output of the KAN positive?
Or we should use function getabs to transfer it to positive value?

**Comment by KindXiaoming** [comment]
hi, try set `update_grid=False` in `train`. If this helps, this means there's a potential bug with `update_grid` or your dataset is somewhat singular.

**Comment by FulingChen** [comment]
I had the same issue. It could be due to different reasons. My problem happened because my input data had some categorical features, which made the spline struggled to find the estimation on the grid. My problem happend when it is on torch.linalg.lstsq, because it cannot solve A = XB. The method @KindXiaoming suggested could be helpful but if you do need to update the grid, you might need to add the regulation in your spline code to get ride of the zeros of A and B. Here is what I have and it fixed my problem.

In spline.py, function curve2coef(x_eval, y_eval, grid, k, device="cpu"), add this 

    # # coef = torch.linalg.lstsq(mat, y_eval.unsqueeze(dim=2)).solution[:, :, 0]
    #
    # coef = torch.linalg.lstsq(mat.to(device), y_eval.unsqueeze(dim=2).to(device),
    #                           driver='gelsy' if device == 'cpu' else 'gels').solution[:, :, 0]

    lambda_reg = 1e-5
    batch_size, rows, cols = mat.shape

    # Placeholder for coefficients
    coef = torch.zeros((batch_size, cols), device=device)

    # Iterate over each batch
    for i in range(batch_size):
        # Extract the i-th batch
        mat_batch = mat[i].to(device)
        y_eval_batch = y_eval[i].to(device)

        # Regularization
        I = torch.eye(mat_batch.size(1), device=device)  # Identity matrix
        mat_reg = mat_batch.T @ mat_batch + lambda_reg * I
        y_reg = mat_batch.T @ y_eval_batch.unsqueeze(dim=1)

        # Perform least squares computation with exception handling
        try:
            batch_coef = torch.linalg.lstsq(mat_reg, y_reg, driver='gelsy' if device == 'cpu' else 'gels').solution[:, 0]
            coef[i] = batch_coef
        except RuntimeError as e:
            raise RuntimeError(f"Error during least squares computation in batch {i}: {e}")

    return coef.to(device)

**Comment by chen-erqi** [comment]
> hi, try set `update_grid=False` in `train`. If this helps, this means there's a potential bug with `update_grid_from_samples` or your dataset is somewhat singular.

hi! Thanks for your amazing multiKan.
But unfortunately I had the "nan" question during training.
And solve it by changing update grid to False.
And that is weired why it can be effect?
The KANLayer.forward is used with or without update grid.
Could you please to discuss this question with me? Thanks a lot!
